src/api/substring.cc

```diff
@@ -0,0 +1,130 @@
+
+    auto transaction = GetTransaction(transactionId);
+    if (transaction == nullptr) {
+        args.GetReturnValue().Set(vResult);
+        return;
+    }
+
+    try {
+        uintptr_t subjectPointer = utils::GetLocalStringPointer(vSubject);
+        auto taintedObj = transaction->FindTaintedObject(subjectPointer);
+        auto oRanges = taintedObj ? taintedObj->getRanges() : nullptr;
+
+        int substringStart = vSubstringStart->ToInteger(context).ToLocalChecked()->Value();
+        int substringEnd = argc > 4 ?
+            vSubstringEnd->ToInteger(context).ToLocalChecked()->Value() :
+            vSubject->ToString(context).ToLocalChecked()->Length();
+
+        Local<String> result = vResult->ToString(context).ToLocalChecked();
+
+        if (oRanges != nullptr) {
+            auto newRanges = transaction->GetSharedVectorRange();
+            for (auto it = oRanges->begin(); it != oRanges->end(); ++it) {
+                auto oRange = *it;
+                int rangeEnd = oRange->end - substringStart;
+                if (rangeEnd > result->Length()) {
+                    rangeEnd = result->Length();
+                }
+
+                if (substringStart > oRange->start && substringStart < oRange->end) {
+                    if (substringStart != 0 || rangeEnd != oRange->end) {
+                        auto newRange = transaction->GetRange(0, rangeEnd, oRange->inputInfo);
+                        if (newRange) {
+                            newRanges->PushBack(newRange);
+                        } else {
+                            break;
+                        }
+                    } else {
+                        newRanges->PushBack(oRange);
+                    }
+                } else {
+                    if (substringEnd > oRange->start && substringStart < oRange->end) {
+                        if (substringStart != 0 || rangeEnd != oRange->end) {
+                            int rangeStart = oRange->start - substringStart;
+                            auto newRange = transaction->GetRange(rangeStart,
+                                    rangeEnd,
+                                    oRange->inputInfo);
+                            if (newRange) {
+                                newRanges->PushBack(newRange);
+                            } else {
+                                break;
+                            }
+                        } else {
+                            newRanges->PushBack(oRange);
+                        }
+                    }
+                }
+            }
+
+            if (newRanges->Size()) {
+                auto key = utils::GetLocalStringPointer(vResult);
+                transaction->AddTainted(key, newRanges, vResult);
+            }
+        }
+    } catch (const std::bad_alloc& err) {
+    } catch (const container::QueuedPoolBadAlloc& err) {
+    } catch (const container::PoolBadAlloc& err) {
+    }
+
+    args.GetReturnValue().Set(vResult);
```

Comment:
When length of the result is 1, and it is tainted, we need a new external string instance.